# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)
data

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1    4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  \
0            253875                 покупка жилья                     B   
1            112080       приобретение автомобиля                     C   
2            145885                 покупка жилья                     C   
3            267628    дополнительное образование                     B   
4            158616               сыграть свадьбу                     C   
...             ...                           ...                   ...   
21520        224791             операции с жильем                     B   
21521        155999          сделка с автомобилем                     C   
21522         89672                  недвижимость                     C   
21523        244093  на покупку своего автомобиля                     B   
21524         82047         на покупку автомобиля                     C   

               purpose_category  
0      операции с недвижимостью  
1        операции с автомобилем  
2      операции с недвижимостью  
3         получение образования  
4            проведение свадьбы  
...                         ...  
21520  операции с недвижимостью  
21521    операции с автомобилем  
21522  операции с недвижимостью  
21523    операции с автомобилем  
21524    операции с автомобилем  

[21331 rows x 14 columns]

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
data_children_debt = pd.DataFrame()
data_children_debt['sum_children'] = data.groupby('children')['debt'].sum()
data_children_debt['count_children'] = data.groupby('children')['debt'].count()
data_children_debt['%'] = (data_children_debt['sum_children'] / data_children_debt['count_children'])*100 
data_children_debt.sort_values('%', ascending = False)

sum_children  count_children         %
children                                        
4                    4              41  9.756098
2                  194            2052  9.454191
1                  444            4808  9.234609
3                   27             330  8.181818
0                 1063           14091  7.543822
5                    0               9  0.000000

**Вывод: Из полученных данных видим, что категория "Бездетные" чаще оплачивает кредит в срок (7.54% просрочек), нежели категории с детьми (8.18% и выше). В группе "5 детей" мы видим, что количетсво просрочек равно 0. Было бы ошибочно делать вывод, что такие семьи платят без задержек, так как мы видим, что семьи с детьми чаще имеют просрочки.** 

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
data_family_status_debt = pd.DataFrame()
data_family_status_debt['sum_family_status'] = data.groupby('family_status')['debt'].sum()
data_family_status_debt['count_family_status'] = data.groupby('family_status')['debt'].count()
data_family_status_debt['%'] = (data_family_status_debt['sum_family_status'] / data_family_status_debt['count_family_status'])*100 
data_family_status_debt.sort_values('%', ascending = False)

sum_family_status  count_family_status         %
family_status                                                          
Не женат / не замужем                273                 2796  9.763948
гражданский брак                     385                 4134  9.313014
женат / замужем                      927                12261  7.560558
в разводе                             84                 1189  7.064760
вдовец / вдова                        63                  951  6.624606

**Вывод: Из полученных данных видим, что люди со статусом "Не женат / не замужем" и "В гражданском браке" менее ответсвенны и чаще задерживают оплату по кредиту, относительно людей, состоящих/бывших в браке. Интересная зависимость, так как часто люди в браке имеют детей. А из прошлого наблюдения мы заметили, что бездетные лучше платят, чем с детьми.** 

In [24]:
# Проверил свое утверждение "часто люди в браке имеют детей".
def child(row):
    children = row['children']
    family_status = row['family_status']
    if children > 0:
        if family_status == 'женат / замужем':
            return 'с детьми'
    if children > 0:
        if family_status == 'в разводе':
            return 'с детьми'
    if children > 0:
        if family_status == 'вдовец / вдова':
            return 'с детьми'
    if children == 0:
        if family_status == 'женат / замужем':
            return 'без детей'
    if children == 0:
        if family_status == 'в разводе':
            return 'без детей'
    if children == 0:
        if family_status == 'вдовец / вдова':
            return 'без детей'
    return 'не учитываем'
data['children_status'] = data.apply(child, axis=1)
data.iloc[:,[0, 5, 14]].head(40)

children          family_status children_status
0          1        женат / замужем        с детьми
1          1        женат / замужем        с детьми
2          0        женат / замужем       без детей
3          3        женат / замужем        с детьми
4          0       гражданский брак    не учитываем
5          0       гражданский брак    не учитываем
6          0        женат / замужем       без детей
7          0        женат / замужем       без детей
8          2       гражданский брак    не учитываем
9          0        женат / замужем       без детей
10         2        женат / замужем        с детьми
11         0        женат / замужем       без детей
12         0       гражданский брак    не учитываем
13         0        женат / замужем       без детей
14         0       гражданский брак    не учитываем
15         1        женат / замужем        с детьми
16         0        женат / замужем       без детей
17         0       гражданский брак    не учитываем
18         0         вдовец / вдова       без детей
19         0              в разводе       без детей
20         1        женат / замужем        с детьми
21         1       гражданский брак    не учитываем
22         1       гражданский брак    не учитываем
23         0       гражданский брак    не учитываем
24         1  Не женат / не замужем    не учитываем
25         0        женат / замужем       без детей
26         0        женат / замужем       без детей
27         0        женат / замужем       без детей
28         1        женат / замужем        с детьми
29         0  Не женат / не замужем    не учитываем
30         1        женат / замужем        с детьми
31         0        женат / замужем       без детей
32         0       гражданский брак    не учитываем
33         0        женат / замужем       без детей
34         0        женат / замужем       без детей
35         0       гражданский брак    не учитываем
36         0  Не женат / не замужем    не учитываем
37         0       гражданский брак    не учитываем
38         0  Не женат / не замужем    не учитываем
39         0       гражданский брак    не учитываем

In [25]:
data.groupby('children_status')['children'].count().sort_values(ascending = False)

children_status
без детей       9099
не учитываем    6930
с детьми        5302
Name: children, dtype: int64

Вижу, что я со своим выводом поторопился. \
Оказалось, что клиентов в ***браке/бывших в браке с детьми*** практически **вдвое меньше**, чем клиентов ***в браке/бывших в браке без детей***.

In [26]:
data_children_status_debt = pd.DataFrame()
data_children_status_debt['sum_children_status'] = data.groupby('children_status')['debt'].sum()
data_children_status_debt['count_children_status'] = data.groupby('children_status')['debt'].count()
data_children_status_debt['%'] = (data_children_status_debt['sum_children_status'] / data_children_status_debt['count_children_status'])*100 
data_children_status_debt.sort_values('%', ascending = False)

sum_children_status  count_children_status         %
children_status                                                      
не учитываем                     658                   6930  9.494949
с детьми                         450                   5302  8.487363
без детей                        624                   9099  6.857896

**Вывод 2: И все таки подтверждается гипотеза, что бездетные платят исправнее!!!**

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [27]:
# Ранее мы проводили группировку людей по уровню их дохода.
data_total_income_category_debt = pd.DataFrame()
data_total_income_category_debt['sum_total_income_category'] = data.groupby('total_income_category')['debt'].sum()
data_total_income_category_debt['count_total_income_category'] = data.groupby('total_income_category')['debt'].count()
data_total_income_category_debt['%'] = (data_total_income_category_debt['sum_total_income_category'] / data_total_income_category_debt['count_total_income_category'])*100 
data_total_income_category_debt.sort_values('%', ascending = False)

sum_total_income_category  count_total_income_category  \
total_income_category                                                           
E                                              2                           22   
C                                           1353                        15921   
A                                              2                           25   
B                                            354                         5014   
D                                             21                          349   

                              %  
total_income_category            
E                      9.090909  
C                      8.498210  
A                      8.000000  
B                      7.060231  
D                      6.017192

**Вывод: Из полученных данных мы наблюдаем необычное явление, люди с высоким доходом (A) платят хуже, чем люди с доходом ниже среднего (D). На это можно обратить внимание, но, ввиду того, что данных в категориях A, D, E недостаточно, мы не можем брать их в расчет.
Рассматривая категории B и C мы видим, что с увеличением роста доходов, просрочки по выплатам сокращаются.** 

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [28]:
data_purpose_category_debt = pd.DataFrame()
data_purpose_category_debt['sum_purpose_category'] = data.groupby('purpose_category')['debt'].sum()
data_purpose_category_debt['count_purpose_category'] = data.groupby('purpose_category')['debt'].count()
data_purpose_category_debt['%'] = (data_purpose_category_debt['sum_purpose_category'] / data_purpose_category_debt['count_purpose_category'])*100 
data_purpose_category_debt.sort_values('%', ascending = False)

sum_purpose_category  count_purpose_category  \
purpose_category                                                         
операции с автомобилем                     400                    4279   
получение образования                      369                    3988   
проведение свадьбы                         183                    2313   
операции с недвижимостью                   780                   10751   

                                 %  
purpose_category                    
операции с автомобилем    9.347978  
получение образования     9.252758  
проведение свадьбы        7.911803  
операции с недвижимостью  7.255139

**Вывод: Из полученных данных мы видим, что чаще всего просрочка идет по кредитам на авто и образование. Реже на недвижимость и свадьбы.** 

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ: Люди стараются скрыть свой реальный трудовой стаж и уровень дохода, с целью повышения шанса одобрения кредита. \
Возможно, поля "стаж" и "доход" не отмечены пометкой "обязательно к заполнению" и зачастую люди не стремятся заполнять такие поля. Нужно обозначить данный вопрос в отдел разработки, так как пропуск таких важных данных может привести к выдаче кредита не платежеспособным клиентам, либо наоборот, **упущению потенциального клиента и прибыли**.* 

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ: Заполнение пропусков средним значением может привести к искажению данных, так как данные будут усредняться с учетом максимумов и минимумов заданного ряда чисел. \
Если заполнить максиальным или минимальным значением - это также приведет к искажению данных.* 

### Шаг 4: общий вывод.

Проведя анализ полученных данных, можно сделать следующий вывод:
1. Бездетные чаще платят в срок, так как не обеменены расходами на детей.
2. Но в тоже время люди, состоящие/бывшие в браке (вероятнее всего имеют детей) платят исправнее.
3. С ростом доходов платежеспособность клиентов возрастает и количество просрочек сокращается.
4. Кредиты на недвижимость пользуются популярностью и просрочек в данной категории меньше всех. \

**Самые предпочтительные клиенты:** В браке, бездетные, доход выше среднего, оформляющие кредит на недвижимость. Вероятнее всего, объясняется это тем, что в большинсве семей бюджет формируется из суммарного дохода супругов и выплаты соответсвенно идут оттуда.

**Самые нежелательные клиенты:** Одинокие (и не бывшие в браке), с детьми, доходы ниже среднего, оформляющие кредит на авто/учебу.

Для получения более детальных данных, необходимо провести анализ, с разбивкой на категории по возрасту: "молодые", "зрелые", "пенсионеры".